specs/01-sharded-archiving.md

```diff
@@ -0,0 +1,247 @@
+  include both, local segment commitments, and segment commitments from child shards (through
+  `SuperSegment`s). Each of them should be accordingly tagged as such.
+- `shard_blocks[]`: hash map of blocks committed from child shards. It stores the history of the
+  shard in the parent, until these blocks are made available in a new segment.
+- `recorded_history_buffer`: Buffer storing segments (i.e. the local history) of a local shard
+
+### Preparing the history of a shard
+
+#### 1. Genesis Block Archiving
+
+- Input: Genesis block of the shard.
+- The genesis block is archived as soon as it is produced. We extend the encoding of the genesis
+  block with extra pseudorandom data up to `RECORDED_HISTORY_SEGMENT_SIZE`, such that the very first
+  archived segment can be produced right away, bootstrapping the farming process. This extra data is
+  added to the end of the SCALE-encoded block, so during decoding of the genesis block it'll be
+  ignored.
+- This becomes the first segment of the shard's history.
+- Output: First archived segment.
+
+#### 2. Block History Buffering
+
+- Input: Blocks produced in the shard.
+- Once any block produced after genesis becomes `CONFIRMATION_DEPTH_K`-deep, it is included in the
+  recorded history. New blocks are added to recorded history that is a buffer of capacity
+  `RECORDED_HISTORY_SEGMENT_SIZE`.
+- When added to the buffer, blocks are turned into `SegmentItem`s.
+- Each segment will have the parent `segment_header` included as the first item. Each
+  `segment_header` includes the hash of the previous segment header and the segment commitment of
+  the previous segment. Together, segment headers form a chain that is used for quick and efficient
+  verification of pieces corresponding to the actual archival history of the blockchain.
+- Segment items are combined into segments. Each segment contains at least two of the following
+  `SegmentItem`s, in this order:
+- The previous segment’s SegmentHeader
+  - BlockContinuation, if remained from the previous segment
+  - Block(s), as many as fit fully into the current segment, may be none
+  - BlockStart, if the next block doesn't fit within the current segment
+  - Padding (zero or more) in case padding bytes are necessary to complete the segment to
+    `RECORDED_HISTORY_SEGMENT_SIZE`
+- Output: Buffered SegmentItems.
+
+#### 3. Block Info Construction
+
+- Input: Blocks produced in the shard.
+- When a new block is produced and included to the chain in a shard, farmers (and any other node in
+  the system), extend the block using an error correction scheme to generate parity shares:
+  1. The block is split into `M = ceil(block_size/l)` shares of `l` bytes.
+  2. Each share is arranged into a different column of a `M x l = ceil(block_size/l) x l` matrix.
+  3. Shares are erasure coded column-wise with `extend(column, ERASURE_CODING_RATE)`. This
+     effectively doubles the number of rows and, hence, the shares per block
+     to `total_shares = 2*M`.
```

Comment:
This doesn't really compute in my head in a straightforward way.

When I think of a block sliced into shared that are then erasure coded, I imagine this:

| column_0 | column_1 |
|-|-|
| source_share_0 | source_share_1 |

Turned into this:

| column_0 | column_1 | column_2 | column_3 |
|-|-|-|-|
| source_share_0 | source_share_1 | parity_share_0 | parity_share_1 |

With `source_share0 || source_share1` being the block (+padding).

But you clearly meant something different since you're talking about doubling the number of rows.